src/config.js

```diff
@@ -0,0 +1,1602 @@
+// ```
+//
+// In other words, if a function asks for a `key-path`, that path doesn't have to
+// be described in the config schema for the package or any package. However, as
+// highlighted in the best practices section, you are advised against doing the
+// above.
+//
+// ## Best practices
+//
+// * Don't depend on (or write to) configuration keys outside of your keypath.
+//
+const schemaEnforcers = {}
+
+class Config {
+  static addSchemaEnforcer (typeName, enforcerFunction) {
+    if (schemaEnforcers[typeName] == null) { schemaEnforcers[typeName] = [] }
+    return schemaEnforcers[typeName].push(enforcerFunction)
+  }
+
+  static addSchemaEnforcers (filters) {
+    for (let typeName in filters) {
+      const functions = filters[typeName]
+      for (let name in functions) {
+        const enforcerFunction = functions[name]
+        this.addSchemaEnforcer(typeName, enforcerFunction)
+      }
+    }
+  }
+
+  static executeSchemaEnforcers (keyPath, value, schema) {
+    let error = null
+    let types = schema.type
+    if (!Array.isArray(types)) { types = [types] }
+    for (let type of types) {
+      try {
+        const enforcerFunctions = schemaEnforcers[type].concat(schemaEnforcers['*'])
+        for (let enforcer of Array.from(enforcerFunctions)) {
+            // At some point in one's life, one must call upon an enforcer.
+          value = enforcer.call(this, keyPath, value, schema)
+        }
+        error = null
+        break
+      } catch (e) {
+        error = e
+      }
+    }
+
+    if (error != null) { throw error }
+    return value
+  }
+
+  // Created during initialization, available as `atom.config`
+  constructor (param = {}) {
+    const {notificationManager, enablePersistence} = param
+    this.notificationManager = notificationManager
+    this.enablePersistence = enablePersistence
+    this.clear()
+  }
+
+  initialize ({configDirPath, resourcePath, projectHomeSchema}) {
+    this.configDirPath = configDirPath
+    this.resourcePath = resourcePath
+    if (this.enablePersistence != null) {
+      this.configFilePath = fs.resolve(this.configDirPath, 'config', ['json', 'cson'])
+      if (this.configFilePath == null) { this.configFilePath = path.join(this.configDirPath, 'config.cson') }
+    }
+
+    this.schema.properties.core.properties.projectHome = projectHomeSchema
+    this.defaultSettings.core.projectHome = projectHomeSchema.default
+  }
+
+  clear () {
+    this.emitter = new Emitter()
+    this.schema = {
+      type: 'object',
+      properties: {}
+    }
+    this.defaultSettings = {}
+    this.settings = {}
+    this.scopedSettingsStore = new ScopedPropertyStore()
+
+    this.settingsLoaded = false
+    this.savePending = false
+    this.configFileHasErrors = false
+    this.transactDepth = 0
+    this.pendingOperations = []
+    this.legacyScopeAliases = {}
+
+    this.requestLoad = _.debounce(() => {
+      this.loadUserConfig()
+    }
+      , 100)
```

Comment:
the indentation here is pretty weird. maybe this?

```js
_.debounce(
  () => { this.loadUserConfig() },
  100
)
```

Reply:
^